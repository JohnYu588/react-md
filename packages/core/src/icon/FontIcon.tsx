import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";

import type { FontIconClassNameOptions } from "./styles";
import { icon } from "./styles";

export interface FontIconProps
  extends Omit<HTMLAttributes<HTMLElement>, "color">,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 *
 * \@remarks \@since 6.0.0 Switched from `<i>` to `<span>` element.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(props, ref) {
    const {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      color,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...remaining
    } = props;

    return (
      <span
        {...remaining}
        aria-hidden={ariaHidden}
        ref={ref}
        className={icon({
          type: "font",
          dense,
          color,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </span>
    );
  }
);
