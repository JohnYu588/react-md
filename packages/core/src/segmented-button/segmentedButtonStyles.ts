import { cnb } from "cnbuilder";

import { cssUtils } from "../cssUtils.js";
import { bem } from "../utils/bem.js";

const styles = bem("rmd-segmented-button");

/**
 * @since 6.0.0
 */
export interface SegmentedButtonClassNameOptions {
  className?: string;

  /** @defaultValue `false` */
  selected?: boolean;

  /**
   * An optional className to apply when {@link selected} is `true`
   */
  selectedClassName?: string;

  /** @internal */
  pressedClassName?: string;
}

/**
 * @since 6.0.0
 */
export function segmentedButton(
  options: SegmentedButtonClassNameOptions = {}
): string {
  const { className, selected, selectedClassName, pressedClassName } = options;

  return cnb(
    styles({ selected }),
    selected && selectedClassName,
    pressedClassName,
    cssUtils({ surface: true }),
    className
  );
}
