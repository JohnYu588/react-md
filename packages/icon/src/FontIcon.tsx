import type { HTMLAttributes, ReactNode } from "react";
import { forwardRef } from "react";
import type { FontIconClassNameOptions } from "./getIconClassName";
import { getIconClassName } from "./getIconClassName";

export interface FontIconProps
  extends HTMLAttributes<HTMLElement>,
    FontIconClassNameOptions {
  /**
   * Any children to render to create the font icon. This is required for
   * material-icons.
   */
  children?: ReactNode;
}

/**
 * The `FontIcon` component is used for rendering a font-icon library's icon.
 * The default is to use the `material-icons` library, but others can be used as
 * well.
 *
 * If you are using another font icon library that does not always create icons
 * with a perfect 1:1 scale (such as font awesome), it is recommended to use the
 * `forceSize` and `forceFontSize` props to fix the sizing issues.
 */
export const FontIcon = forwardRef<HTMLElement, FontIconProps>(
  function FontIcon(
    {
      className,
      children,
      "aria-hidden": ariaHidden = true,
      dense = false,
      iconClassName = "material-icons",
      forceSize = false,
      forceFontSize = false,
      ...props
    },
    ref
  ) {
    return (
      <i
        {...props}
        aria-hidden={ariaHidden}
        ref={ref}
        className={getIconClassName({
          type: "font",
          dense,
          className,
          iconClassName,
          forceSize,
          forceFontSize,
        })}
      >
        {children}
      </i>
    );
  }
);
