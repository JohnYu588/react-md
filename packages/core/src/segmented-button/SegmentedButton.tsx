"use client";

import { type ButtonHTMLAttributes, type ReactNode, forwardRef } from "react";

import { getIcon } from "../icon/iconConfig.js";
import { type ComponentWithRippleProps } from "../interaction/types.js";
import { useElementInteraction } from "../interaction/useElementInteraction.js";
import { useHigherContrastChildren } from "../interaction/useHigherContrastChildren.js";
import { useMaxWidthTransition } from "../transition/useMaxWidthTransition.js";
import { segmentedButton } from "./segmentedButtonStyles.js";

/**
 * @since 6.0.0
 */
export interface SegmentedButtonProps
  extends ButtonHTMLAttributes<HTMLButtonElement>,
    ComponentWithRippleProps {
  /**
   * Set this to `true` to apply selected styles and an optional
   * {@link selectedIcon}
   *
   * @defaultValue `false`
   */
  selected?: boolean;

  /** @defaultValue `getIcon("selected")` */
  selectedIcon?: ReactNode;

  /**
   * An optional className to apply when {@link selected} is `true`.
   */
  selectedClassName?: string;

  /**
   * Set this to `true` to not render the {@link selectedIcon} when
   * {@link selected} is `true`.
   *
   * @defaultValue `false`
   */
  disableSelectedIcon?: boolean;

  /**
   * Set this to `true` to disable the {@link selectedIcon} enter/exit
   * transition and instead just use `display: none`.
   *
   * @defaultValue `false`
   */
  disableSelectedTransition?: boolean;

  /**
   * An optional addon to render before the {@link children} and after the
   * {@link selectedIcon}. This is only useful when rendering text children so
   * it can appear above the interaction states.
   */
  leftAddon?: ReactNode;

  /**
   * An optional addon to render after the {@link children}. This is only useful
   * when rendering text children so it can appear above the interaction states.
   */
  rightAddon?: ReactNode;
}

/**
 * **Client Component**
 *
 * @example Simple Example
 * ```tsx
 * import {
 *   SegmentedButton,
 *   SegmentedButtonContainer,
 * } from "@react-md/core";
 * import type { ReactElement } from "react";
 * import { useState } from "react";
 *
 * function Example(): ReactElement {
 *   const [value, setValue] = useState("a");
 *   return (
 *     <SegmentedButtonContainer>
 *       <SegmentedButton
 *         onClick={() => setValue("a")}
 *         selected={value === "a"}
 *       >
 *         First
 *       </SegmentedButton>
 *       <SegmentedButton
 *         onClick={() => setValue("b")}
 *         selected={value === "b"}
 *       >
 *         Second
 *       </SegmentedButton>
 *       <SegmentedButton
 *         onClick={() => setValue("c")}
 *         selected={value === "c"}
 *         disableSelectedIcon
 *       >
 *         Third
 *       </SegmentedButton>
 *     </SegmentedButtonContainer>
 *   );
 * }
 * ```
 *
 * @since 6.0.0
 */
export const SegmentedButton = forwardRef<
  HTMLButtonElement,
  SegmentedButtonProps
>(function SegmentedButton(props, ref) {
  const {
    className,
    type = "button",
    leftAddon,
    rightAddon,
    children: propChildren,
    selected,
    selectedIcon: propSelectedIcon,
    selectedClassName,
    disableSelectedIcon,
    disableSelectedTransition,
    onBlur,
    onClick,
    onKeyDown,
    onKeyUp,
    onMouseDown,
    onMouseUp,
    onMouseLeave,
    onDragStart,
    onTouchStart,
    onTouchEnd,
    onTouchMove,
    disabled,
    disableRipple,
    ...remaining
  } = props;

  const children = useHigherContrastChildren(propChildren);
  const selectedIconNode = getIcon("selected", propSelectedIcon);
  const selectedIcon = useMaxWidthTransition({
    element: selectedIconNode,
    transitionIn: !!selected,
    disabled: disableSelectedIcon,
    disableTransition: disableSelectedTransition,
  });
  const { pressedClassName, handlers, ripples } = useElementInteraction({
    mode: disableRipple ? "none" : undefined,
    onBlur,
    onClick,
    onKeyDown,
    onKeyUp,
    onMouseDown,
    onMouseUp,
    onMouseLeave,
    onDragStart,
    onTouchStart,
    onTouchEnd,
    onTouchMove,
    disabled,
  });

  return (
    <button
      {...remaining}
      {...handlers}
      aria-pressed={selected}
      ref={ref}
      type={type}
      disabled={disabled}
      className={segmentedButton({
        className,
        selected,
        selectedClassName,
        pressedClassName,
      })}
    >
      {!disableSelectedIcon && selectedIcon}
      {leftAddon}
      {children}
      {rightAddon}
      {ripples}
    </button>
  );
});
