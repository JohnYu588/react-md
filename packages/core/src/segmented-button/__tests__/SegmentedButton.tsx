import { describe, expect, it } from "@jest/globals";
import { createRef } from "react";
import { FontIcon } from "../../icon/FontIcon.js";
import { rmdRender, screen } from "../../test-utils/index.js";
import { SegmentedButton } from "../SegmentedButton.js";
import { segmentedButton } from "../segmentedButtonStyles.js";

describe("SegmentedButton", () => {
  it("should apply the correct styling, HTMLAttributes, and allow a ref", () => {
    const ref = createRef<HTMLButtonElement>();
    const props = {
      ref,
      children: "Button",
    } as const;

    const { rerender } = rmdRender(<SegmentedButton {...props} />);

    const element = screen.getByRole("button", { name: "Button" });
    expect(ref.current).toBeInstanceOf(HTMLButtonElement);
    expect(ref.current).toBe(element);
    expect(element).toMatchSnapshot();

    rerender(
      <SegmentedButton
        {...props}
        style={{ color: "white" }}
        className="custom-class-name"
      />
    );
    expect(element).toMatchSnapshot();

    rerender(<SegmentedButton {...props} selected />);
    expect(element).toMatchSnapshot();

    rerender(<SegmentedButton {...props} disableSelectedIcon />);
    expect(element).toMatchSnapshot();

    rerender(<SegmentedButton {...props} disableSelectedTransition />);
    expect(element).toMatchSnapshot();
  });

  it("should support rendering left and right addons so the children can be text and elevated", () => {
    rmdRender(
      <SegmentedButton
        leftAddon={<FontIcon>home</FontIcon>}
        rightAddon={<FontIcon>check</FontIcon>}
      >
        Content
      </SegmentedButton>
    );

    const button = screen.getByRole("button", { name: "Content" });
    expect(button).toMatchSnapshot();
  });

  describe("styling utility class", () => {
    it("should be callable without any arguments", () => {
      expect(segmentedButton()).toMatchSnapshot();
    });
  });
});
