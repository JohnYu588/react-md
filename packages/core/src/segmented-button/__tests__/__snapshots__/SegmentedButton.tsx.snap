// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SegmentedButton should apply the correct styling, HTMLAttributes, and allow a ref 1`] = `
<button
  class="rmd-segmented-button rmd-interaction-surface"
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-max-width-transition"
  >
    check
  </span>
  <span>
    Button
  </span>
</button>
`;

exports[`SegmentedButton should apply the correct styling, HTMLAttributes, and allow a ref 2`] = `
<button
  class="rmd-segmented-button rmd-interaction-surface custom-class-name"
  style="color: white;"
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-max-width-transition"
  >
    check
  </span>
  <span>
    Button
  </span>
</button>
`;

exports[`SegmentedButton should apply the correct styling, HTMLAttributes, and allow a ref 3`] = `
<button
  aria-pressed="true"
  class="rmd-segmented-button rmd-segmented-button--selected rmd-interaction-surface"
  style=""
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-max-width-transition rmd-max-width-transition--visible"
  >
    check
  </span>
  <span>
    Button
  </span>
</button>
`;

exports[`SegmentedButton should apply the correct styling, HTMLAttributes, and allow a ref 4`] = `
<button
  class="rmd-segmented-button rmd-interaction-surface"
  style=""
  type="button"
>
  <span>
    Button
  </span>
</button>
`;

exports[`SegmentedButton should apply the correct styling, HTMLAttributes, and allow a ref 5`] = `
<button
  class="rmd-segmented-button rmd-interaction-surface"
  style=""
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-display-none"
  >
    check
  </span>
  <span>
    Button
  </span>
</button>
`;

exports[`SegmentedButton should support rendering left and right addons so the children can be text and elevated 1`] = `
<button
  class="rmd-segmented-button rmd-interaction-surface"
  type="button"
>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons rmd-max-width-transition"
  >
    check
  </span>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    home
  </span>
  <span>
    Content
  </span>
  <span
    aria-hidden="true"
    class="rmd-icon rmd-icon--font material-icons"
  >
    check
  </span>
</button>
`;

exports[`SegmentedButton styling utility class should be callable without any arguments 1`] = `"rmd-segmented-button rmd-interaction-surface"`;
